src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java

```diff
@@ -0,0 +1,154 @@
+import static org.apache.commons.lang3.StringUtils.stripToNull;
+
+/**
+ * The details of an environment that was deployed to.
+ *
+ * @since deployments
+ */
+public final class BitbucketDeploymentEnvironment {
+
+    private static final String DISPLAY_NAME = "displayName";
+    private static final String KEY = "key";
+    private static final String TYPE = "type";
+    private static final String URL = "url";
+
+    private final String key;
+    private final String name;
+    private final String type;
+    private final String url;
+
+    private BitbucketDeploymentEnvironment(Builder builder) {
+        key = builder.key;
+        name = builder.name;
+        type = builder.type == null ? null : builder.type.name();
+        url = builder.url;
+    }
+
+    @JsonCreator
+    public BitbucketDeploymentEnvironment(@JsonProperty(KEY) String key,
+                                          @JsonProperty(DISPLAY_NAME) String name,
+                                          @JsonProperty(TYPE) String type,
+                                          @JsonProperty(URL) String url) {
+        this.key = key;
+        this.name = name;
+        this.type = type;
+        this.url = url;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        BitbucketDeploymentEnvironment that = (BitbucketDeploymentEnvironment) o;
+        return Objects.equals(key, that.key) && Objects.equals(name, that.name) &&
+               Objects.equals(type, that.type) && Objects.equals(url, that.url);
+    }
+
+    /**
+     * @return a unique identifier for this environment
+     */
+    @JsonProperty(DISPLAY_NAME)
+    @Nonnull
+    public String getKey() {
+        return key;
+    }
+
+    /**
+     * @return a human-readable name for this environment
+     */
+    @JsonProperty(KEY)
+    @Nonnull
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * @return the {@link Type} of environment, or {@code null} to indicate no type
+     */
+    @JsonProperty(TYPE)
+    @Nullable
+    public String getType() {
+        return type;
+    }
+
+    /**
+     * @return the URL of the environment, or {@code null} to indicate no URL
+     */
+    @JsonProperty(URL)
+    @Nullable
+    public String getUrl() {
+        return url;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(key, name, type, url);
+    }
+
+    @Override
+    public String toString() {
+        return "BitbucketDeploymentEnvironment{" +
+               "key='" + key + '\'' +
+               ", name='" + name + '\'' +
+               ", type='" + type + '\'' +
+               ", url='" + url + '\'' +
+               '}';
+    }
+
+    /**
+     * The types of environments available via the Bitbucket Server API.
+     */
+    public enum Type {
```

Comment:
could rename this to be EnvironmentType, Type is a bit generic

Reply:
Yeah, I thought that since it's an inner class of `BitbucketDeploymentEnvironment` it was obvious what 'type' it was. Maybe it's better if I pull it out into it's own class